File: src/testRunner/transpileRunner.ts
import {
    Baseline,
    Compiler,
    getFileBasedTestConfigurations,
    IO,
    RunnerBase,
    TestCaseParser,
    TestRunnerKind,
} from "./_namespaces/Harness.js";
import * as ts from "./_namespaces/ts.js";
import * as vpath from "./_namespaces/vpath.js";

export class TranspileRunner extends RunnerBase {
    protected basePath = "tests/cases/transpile";
    protected testSuiteName: TestRunnerKind = "transpile";

    public enumerateTestFiles(): string[] {
        // see also: `enumerateTestFiles` in tests/webTestServer.ts
        return this.enumerateFiles(this.basePath, /\.[cm]?[tj]sx?/i, { recursive: true });
    }

    public kind(): TestRunnerKind {
        return this.testSuiteName;
    }

    public initializeTests(): void {
        if (this.tests.length === 0) {
            this.tests = IO.enumerateTestFiles(this);
        }

        describe(this.testSuiteName + " tests", () => {
            this.tests.forEach(file => {
                file = vpath.normalizeSeparators(file);
                describe(file, () => {
                    const tests = TranspileTestCase.getConfigurations(file);
                    for (const test of tests) {
                        test.run();
                    }
                });
            });
        });
    }
}

enum TranspileKind {
    Module,
    Declaration,
}

class TranspileTestCase {
    static varyBy = [
        "declarationMap",
        "sourceMap",
        "inlineSourceMap",
    ];

    static getConfigurations(file: string): TranspileTestCase[] {
        const ext = vpath.extname(file);
        const baseName = vpath.basename(file);
        const justName = baseName.slice(0, baseName.length - ext.length);
        const content = IO.readFile(file)!;
        const settings = TestCaseParser.extractCompilerSettings(content);
        const settingConfigurations = getFileBasedTestConfigurations(settings, TranspileTestCase.varyBy);
        return settingConfigurations?.map(c => {
            const desc = Object.entries(c).map(([key, value]) => `${key}=${value}`).join(",");
            return new TranspileTestCase(`${justName}(${desc})`, ext, content, { ...settings, ...c });
        }) ?? [new TranspileTestCase(justName, ext, content, settings)];
    }

    private jsOutName;
    private dtsOutName;
    private units;
    constructor(
        private justName: string,
        private ext: string,
        private content: string,
        private settings: TestCaseParser.CompilerSettings,
    ) {
        this.jsOutName = justName + this.getJsOutputExtension(`${justName}${ext}`);
        this.dtsOutName = justName + ts.getDeclarationEmitExtensionForPath(`${justName}${ext}`);
        this.units = TestCaseParser.makeUnitsFromTest(content, `${justName}${ext}`, settings);
    }

    getJsOutputExtension(name: string) {
        return ts.getOutputExtension(name, { jsx: this.settings.jsx === "preserve" ? ts.JsxEmit.Preserve : undefined });
    }

    runKind(kind: TranspileKind) {
        it(`transpile test ${this.justName} has expected ${kind === TranspileKind.Module ? "js" : "declaration"} output`, () => {
            let baselineText = "";

            // include inputs in output so how the test is parsed and broken down is more obvious
            this.units.testUnitData.forEach(unit => {
                baselineText += `//// [${unit.name}] ////\r\n`;
                baselineText += unit.content;
                if (!unit.content.endsWith("\n")) {
                    baselineText += "\r\n";
                }
            });

            this.units.testUnitData.forEach(unit => {
                const opts: ts.CompilerOptions = {};
                Compiler.setCompilerOptionsFromHarnessSetting(this.settings, opts);
                const result = (kind === TranspileKind.Module ? ts.transpileModule : ts.transpileDeclaration)(unit.content, { compilerOptions: opts, fileName: unit.name, reportDiagnostics: this.settings.reportDiagnostics === "true" });

                baselineText += `//// [${ts.changeExtension(unit.name, kind === TranspileKind.Module ? this.getJsOutputExtension(unit.name) : ts.getDeclarationEmitExtensionForPath(unit.name))}] ////\r\n`;
                baselineText += result.outputText;
                if (!result.outputText.endsWith("\n")) {
                    baselineText += "\r\n";
                }
                if (result.sourceMapText) {
                    baselineText += `//// [${ts.changeExtension(unit.name, kind === TranspileKind.Module ? this.getJsOutputExtension(unit.name) : ts.getDeclarationEmitExtensionForPath(unit.name))}.map] ////\r\n`;
                    baselineText += result.sourceMapText;
                    if (!result.outputText.endsWith("\n")) {
                        baselineText += "\r\n";
                    }
                }
                if (result.diagnostics && result.diagnostics.length) {
                    baselineText += "\r\n\r\n//// [Diagnostics reported]\r\n";
                    baselineText += Compiler.getErrorBaseline([{ content: unit.content, unitName: unit.name }], result.diagnostics, !!opts.pretty);
                    if (!baselineText.endsWith("\n")) {
                        baselineText += "\r\n";
                    }
                }
            });

            Baseline.runBaseline(`transpile/${kind === TranspileKind.Module ? this.jsOutName : this.dtsOutName}`, baselineText);
        });
    }

    run() {
        if (!this.settings.emitDeclarationOnly) {
            this.runKind(TranspileKind.Module);
        }
        if (this.settings.declaration) {
            this.runKind(TranspileKind.Declaration);
        }
    }
}

File: tsc/internal/testrunner/transpile_runner.go
package testrunner

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"testing"

	"github.com/microsoft/typescript-go/internal/core"
	"github.com/microsoft/typescript-go/internal/diagnosticwriter"
	"github.com/microsoft/typescript-go/internal/outputpaths"
	"github.com/microsoft/typescript-go/internal/repo"
	"github.com/microsoft/typescript-go/internal/testutil/baseline"
	"github.com/microsoft/typescript-go/internal/testutil/harnessutil"
	"github.com/microsoft/typescript-go/internal/testutil/tsbaseline"
	"github.com/microsoft/typescript-go/internal/transpile"
	"github.com/microsoft/typescript-go/internal/tspath"
	"github.com/microsoft/typescript-go/internal/vfs/osvfs"
)

var transpileBaselineRegex = regexp.MustCompile(`\.[cm]?[tj]sx?$`)

var transpileVaryBy = map[string]struct{}{
	"declarationmap":  {},
	"sourcemap":       {},
	"inlinesourcemap": {},
}

type TranspileBaselineRunner struct {
	testFiles []string
	basePath  string
}

var _ Runner = (*TranspileBaselineRunner)(nil)

func NewTranspileBaselineRunner() *TranspileBaselineRunner {
	return &TranspileBaselineRunner{
		basePath: "../_submodules/TypeScript/tests/cases/transpile",
	}
}

func (r *TranspileBaselineRunner) EnumerateTestFiles() []string {
	if len(r.testFiles) > 0 {
		return r.testFiles
	}
	files, err := harnessutil.EnumerateFiles(r.basePath, transpileBaselineRegex, true)
	if err != nil {
		panic("Could not read transpile test files: " + err.Error())
	}
	r.testFiles = files
	return files
}

func (r *TranspileBaselineRunner) RunTests(t *testing.T) {
	for _, fileName := range r.EnumerateTestFiles() {
		r.runTest(t, fileName)
	}
}

func (r *TranspileBaselineRunner) runTest(t *testing.T, fileName string) {
	content, ok := osvfs.FS().ReadFile(fileName)
	if !ok {
		panic("Could not read transpile test file: " + fileName)
	}
	settings := extractCompilerSettings(content)
	configurations := harnessutil.GetFileBasedTestConfigurations(t, settings, transpileVaryBy)
	if len(configurations) == 0 {
		configurations = []*harnessutil.NamedTestConfiguration{{Config: settings}}
	}

	extension := tspath.GetAnyExtensionFromPath(fileName, nil, false)
	baseName := tspath.GetBaseFileName(fileName)
	justName := strings.TrimSuffix(baseName, extension)
	units := makeUnitsFromTest(content, baseName).testUnitData

	for _, configuration := range configurations {
		configuredName := justName
		if configuration.Name != "" {
			configuredName += "(" + formatTranspileConfigurationName(configuration.Name) + ")"
		}
		t.Run(configuredName, func(t *testing.T) {
			options := &core.CompilerOptions{}
			harnessOptions := &harnessutil.HarnessOptions{}
			harnessutil.SetOptionsFromTestConfig(t, configuration.Config, options, harnessOptions, srcFolder, false)

			if !options.EmitDeclarationOnly.IsTrue() {
				r.runKind(t, configuredName, extension, units, options, harnessOptions, false)
			}
			if options.Declaration.IsTrue() {
				r.runKind(t, configuredName, extension, units, options, harnessOptions, true)
			}
		})
	}
}

func formatTranspileConfigurationName(name string) string {
	name = strings.ReplaceAll(name, "declarationmap=", "declarationMap=")
	name = strings.ReplaceAll(name, "inlinesourcemap=", "inlineSourceMap=")
	return strings.ReplaceAll(name, "sourcemap=", "sourceMap=")
}

func (r *TranspileBaselineRunner) runKind(
	t *testing.T,
	configuredName string,
	extension string,
	units []*testUnit,
	options *core.CompilerOptions,
	harnessOptions *harnessutil.HarnessOptions,
	declaration bool,
) {
	var result strings.Builder
	for _, unit := range units {
		appendTranspileSection(&result, unit.name, unit.content)
	}

	for _, unit := range units {
		transpileOptions := transpile.Options{
			CompilerOptions:   options,
			FileName:          unit.name,
			ReportDiagnostics: harnessOptions.ReportDiagnostics,
		}
		var output *transpile.Output
		if declaration {
			output = transpile.TranspileDeclaration(t.Context(), unit.content, transpileOptions)
		} else {
			output = transpile.TranspileModule(t.Context(), unit.content, transpileOptions)
		}
		if output == nil {
			t.Fatal("transpilation was canceled")
		}

		outputExtension := outputpaths.GetOutputExtension(unit.name, options.Jsx)
		if declaration {
			outputExtension = tspath.GetDeclarationEmitExtensionForPath(unit.name)
		}
		outputFileName := tspath.ChangeExtension(unit.name, outputExtension)
		appendTranspileSection(&result, outputFileName, output.OutputText)
		if output.SourceMapText != "" {
			appendTranspileSection(&result, outputFileName+".map", output.SourceMapText)
		}
		if len(output.Diagnostics) > 0 {
			result.WriteString("\r\n\r\n//// [Diagnostics reported]\r\n")
			diagnosticFileName := unit.name
			if file := output.Diagnostics[0].File(); file != nil {
				diagnosticFileName = file.FileName()
			}
			errorBaseline := tsbaseline.GetErrorBaseline(
				t,
				[]*harnessutil.TestFile{{UnitName: diagnosticFileName, Content: unit.content}},
				diagnosticwriter.WrapASTDiagnostics(output.Diagnostics),
				diagnosticwriter.CompareASTDiagnostics,
				options.Pretty.IsTrue(),
			)
			result.WriteString(strings.ReplaceAll(errorBaseline, diagnosticFileName, unit.name))
			if !strings.HasSuffix(result.String(), "\n") {
				result.WriteString("\r\n")
			}
		}
	}

	baselineExtension := outputpaths.GetOutputExtension(configuredName+extension, options.Jsx)
	if declaration {
		baselineExtension = tspath.GetDeclarationEmitExtensionForPath(configuredName + extension)
	}
	baselineName := configuredName + baselineExtension
	baseline.Run(t, "transpile/"+baselineName, result.String(), baseline.Options{IsSubmodule: true})
}

func appendTranspileSection(result *strings.Builder, fileName string, content string) {
	fmt.Fprintf(result, "//// [%s] ////\r\n", fileName)
	result.WriteString(content)
	if !strings.HasSuffix(content, "\n") {
		result.WriteString("\r\n")
	}
}

func cleanTranspileBaselines() {
	for _, folder := range []string{"submodule", "submoduleAccepted", "submoduleTriaged"} {
		if err := os.RemoveAll(filepath.Join(localBasePath, folder, "transpile")); err != nil {
			panic("Could not clean up transpile baselines: " + err.Error())
		}
	}
}

func RunTranspileTests(t *testing.T) {
	repo.SkipIfNoTypeScriptSubmodule(t)
	cleanTranspileBaselines()
	NewTranspileBaselineRunner().RunTests(t)
}
